surreal/http.py

```diff
@@ -0,0 +1,124 @@
+limitations under the License.
+"""
+import base64
+import logging
+from typing import Any
+from urllib.parse import quote
+
+from aiohttp import ClientSession
+
+from surreal.errors import SurrealException
+from surreal.utils import load_json
+
+from . import __version__
+
+_log = logging.getLogger(__name__)
+
+
+def encode_basic_auth(username: str, password: str):
+    username_password = f"{quote(username)}:{quote(password)}"
+
+    return f"Basic {base64.b64encode(username_password.encode()).decode()}"
+
+
+# NOTE: This implements HTTP state and requests
+class HTTPClient:
+    """
+    Raw HTTP Client used to wrap surreal.
+
+    Parameters
+    ----------
+    url: :class:`str`
+        The url your surrealdb instance is running on.
+    username: :class:`str`
+        The username to use to authenticate with your instance.
+    password: :class:`str`
+        The password to use to authenticate with your instance.
+    ns: :class:`str`
+        The namespace on your instance to use.
+    db: :class:`str`
+        The database on your instance to use.
+    """
+    def __init__(
+        self,
+        url: str,
+        username: str,
+        password: str,
+        ns: str,
+        db: str,
+    ) -> None:
+        self.url = f"{url}/sql"
+        self._session: ClientSession | None = None
+
+        encoded_auth = encode_basic_auth(username=username, password=password)
+
+        self._base_headers: dict[str, str] = {
+            "Content-Type": "application/json",
+            "Authorization": f"{encoded_auth}",
+            "User-Agent": f"surreal.py ({__version__})",
+            "NS": ns,
+            "DB": db,
+        }
+
+    async def _create_session(self) -> None:
+        self._session = ClientSession()
+
+    async def execute(self, query: str) -> list[dict[str, Any]]:
+        """
+        Executes this query and returns the data given by the server.
+
+        If the query is a transactional query, the list should be ordered based on the transaction order.
+
+        Parameters
+        ----------
+        query: :class:`str`
+            The query to execute.
+
+        Returns
+        -------
+        rows: list[dict[str, Any]]
+        """
+
+        if self._session is None:
+            await self._create_session()
+
+        _log.debug(f"Sending query: {query}")
+
+        r = await self._session.post(self.url, data=query, headers=self._base_headers)
+
+        data: list[dict[str, Any]] = load_json(await r.text())
+
+        _log.debug(f"Got back data: {data}")
+
+        ret: list[dict[str, Any]] = []
+
+        for content in data:
+            if content["status"] == "OK":
+                ret.append(content["result"])
+            elif content["status"] == "ERR":
+                raise SurrealException(
+                    "One of your queries ran into an exception: ",
+                    content.get("detail"),
+                    content.get("information"),
+                )
+
+        return ret
```

Comment:
This does not account for non-2xx status codes. A `400` status code, for example, is returned when the SQL provided is invalid. In this case as far as I can tell this code will return `ret` without an error being raised, which will result in unexpected data in an unknown schema being returned.